### ISSUE TITLE
optimization of bad user agents

### ISSUE THREAD

**Issue by thezoggy** [mitchellkrogza/apache-ultimate-bad-bot-blocker#50]
removed comments to streamline issues for current discussion

**Comment by thezoggy** [comment]
There are a few items you could optimize, however I'm not sure if your matching case sensitive or with a boundary.

example:
if you are using an anchor of `"` to match the start of the user-agent, then `Blow` would take care of `BlowFish`. But `Bandit` would not match for `Web\ Bandit`. Now if you aren't anchoring to `"` and don't have a boundary separator then both these could be optimized for the smallest matching word.

example:
if you are not doing case exact matches, then `Brandprotect` and `BrandProtect` are duplicates.

So I'll just list everything below as I see, depending on your use there are some that should be there.. but also a few that probably shouldn't. Matching 'whole word' of one line within another.

* `Blow` would take care of `BlowFish`
* `Bandit` will already match for `Web\ Bandit`
* `Brandprotect` for `BrandProtect`
* `Collector` for `EMail\ Collector`, `WebImageCollector`, `Web\ Image\ Collector`, `WWW-Collector-E`
* `Copier` for `WebCopier`, `Web\ Copier`
* `CrazyWebCrawler` for `CRAZYWEBCRAWLER`
* `Demon` for `Download\ Demon`
* `Devil` for `Download\ Devil`
* `Disco` for `Discobot`
* `Extractor` for `EMail\ Extractor`, `ExtractorPro`, `LinkextractorPro`, `WebsiteExtractor`, `Website\ Extractor`
* `FrontPage` for `MSFrontPage`
* `Grabber` for `EirGrabber`, `PageGrabber`, `Page\ Grabber`
* `HTTrack` for `WinHTTrack`
* `RankActive` for `RankActiveLinkBot`
* `RankingBot` for `RankingBot2`
* `Reaper` for `WebReaper`, `Web\ Reaper`
~* `Ripper` for `Stripper`, `WebStripper`, `Web\ Stripper`, `Image\ Stripper`~
* `Screaming` for `Screaming\ Frog\ SEO\ Spider`
* `Semrush` for `SemrushBot`
* `SEOkicks` for `SEOkicks-Robot`
* `SISTRIX` for `SISTRIX\ Crawler`
* `Stripper` for `WebStripper`, `Web\ Stripper`, `Image\ Stripper`
* `Teleport` for `TeleportPro`
* `WEBDAV` for `WEBDAV\ Client`
~* `Whack` for `Whacker`, `WebWhacker`, `Web\ Whacker`~
* `Whacker` for `WebWhacker`, `Web\ Whacker`

Now personally, with me using this data for the apache-badbots filter.. 

I do a case insensitive match and anchor the user-agent to the " with a wildcard match afterwards to pick up any/all the variants. So only whole word matches at the start of the line is what I'd want to see optimized..  so that brings the list to just:

* `Blow` would take care of `BlowFish`
* `Brandprotect` for `BrandProtect`
* `CrazyWebCrawler` for `CRAZYWEBCRAWLER`
* `Disco` for `Discobot`
* `RankActive` for `RankActiveLinkBot`
* `RankingBot` for `RankingBot2`
* `Screaming` for `Screaming\ Frog\ SEO\ Spider`
* `Semrush` for `SemrushBot`
* `SEOkicks` for `SEOkicks-Robot`
* `SISTRIX` for `SISTRIX\ Crawler`
* `Teleport` for `TeleportPro`
* `WEBDAV` for `WEBDAV\ Client`

**Comment by mitchellkrogza** [comment]
Thanks @thezoggy I do agree there is some duplications which I will sort out.

The Apache blocker (this repo) uses case insensitive matching with word boundaries as below

```
BrowserMatchNoCase "^(.*?)(\bBlow\b)(.*)$" bad_bot
BrowserMatchNoCase "^(.*?)(\bBlowFish\b)(.*)$" bad_bot
```

And the Nginx Blocker uses the same 

```
"~*\bBlow\b"		3;
"~*\bBlowFish\b" 3;
```

So I will review the list for unnecessary duplications. 👍

**Comment by mitchellkrogza** [comment]
@thezoggy as TravisCI's build system is currently completely down I will only push out the update later today once they have resolved their issues.

**Comment by mitchellkrogza** [comment]
Thanks @thezoggy changes made plus a few other dupes I spotted see https://github.com/mitchellkrogza/apache-ultimate-bad-bot-blocker/commit/90d17c0dbbf3d3aa04d252dadfb4bdde527df6b9

Added additional curl tests into the build process. Thinking `Discoverbot` can also be dropped as it will be matched with `Disco`. Will test offline and confirm.

**Comment by mitchellkrogza** [comment]
@thezoggy also see commit https://github.com/mitchellkrogza/apache-ultimate-bad-bot-blocker/commit/6f71237f1573e66fc9d49a330259ba932a1ad64d have a look and see if you can spot any others and thanks for this contribution 👍

**Comment by mitchellkrogza** [comment]
Re-Added some Bot Names back which were not being detected due to the way I regex match in the blocker. See Commit > https://github.com/mitchellkrogza/apache-ultimate-bad-bot-blocker/commit/8736214316deab02d9afaedb65533265c43661e2

**Comment by mitchellkrogza** [comment]
@thezoggy Have a look at latest list and send me any suggestions for regex changes on matching problems with some of those removals. For instance `Bandit` was positively matched in `Web Bandit` , `Web-Bandit` but not `WebBandit`

**Comment by thezoggy** [comment]
is `Backlink-Ceck` a typo of `backlink-check` or legit (missing h)

**Comment by mitchellkrogza** [comment]
I think it is correct found it on a bad bot list I think, will double check in the morning.

**Comment by mitchellkrogza** [comment]
Is latest list looking a bit better?

**Comment by mitchellkrogza** [comment]
@thezoggy check here

![screen shot 2017-10-31 at 5 43 43 pm](https://user-images.githubusercontent.com/9961541/32233475-26d938e4-be63-11e7-9893-1103e6d3d07e.png)

**Comment by thezoggy** [comment]
double checking the latest list, but so far all looks good

**Comment by mitchellkrogza** [comment]
Great, shout if you find anymore 👍

**Comment by mitchellkrogza** [comment]
@thezoggy welcome to send PR's if you find any more dupes or additions or simply list them here or log them as new issues. I prefer issues system then I can add more tests into the Travis build process to check for errors during the build.

**Comment by thezoggy** [comment]
when you said 'check here' earlier, where is that data? cause I want to look up what `fq` is.. curious if that would result in false positives.. also do you have a list of legit user-strings you compare against to rule out false positives?

**Comment by mitchellkrogza** [comment]
@thezoggy it's here and I am sure if you go through it you can find many more to add. Been meaning to go through it from start to finish but haven't had time and some in their lists are not all bad bots > http://stopbadbots.com/bots-table/?letter=A

**Comment by thezoggy** [comment]
looking though the list, I do not see `fq` on there: http://stopbadbots.com/bots-table/?letter=F

**Comment by thezoggy** [comment]
looking through stopbadbots.com data I see quite a few libs/utilizes that would be blocked if you grabbed all their data. 

for example `python-requests`, `Python-urllib`, `pushbullet`, `node`, `node-superagent` are a few that stand out.. that while deff can be bad but shouldnt be outright blacklisted unless the user opt-in for it.

**Comment by mitchellkrogza** [comment]
Yip there are many that should never be blocked, a lot of Bad Bot lists are badly curated.

**Comment by thezoggy** [comment]
looking at their data for example, their first one is `_zbot`, and if you look at their provided user string:
`Mozilla/5.0 (Windows NT 6.2; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/32.0.1667.0 Safari/537.36 _zbot`

Thankfully I'm updating the apache-badbots to work the same way as yours, using word boundary and matching anywhere in the user agent. Now this does open up false positives.. which is why I'm highly suspect of any pattern that is only a few characters.

So d/ling the latest of that plugin, looking at their asset/bots.txt we see that _zbot is not in their list.. so looks like the data on the website vs whats in their plugin are not in sync.
Their .txt file has 2573 entries, while the website has 2795.

In the .txt they only list the name/nickname/new flag..  so no user agent string sadly. 

Now scraping their site wouldn't be too hard but not something I feel its worth spending any time on.. as the data still needs to be self reviewed.

**Comment by mitchellkrogza** [comment]
@thezoggy I will grab the info out of their Wordpress plugin tomorrow and run the list against all my server logs for 28 sites spanning a year and see what matches I get. This will narrow things down of bots to inspect further before adding any new ones. I always like to make sure a bad bot being added is valid.

**Comment by thezoggy** [comment]
made a quick excel sheet, loaded in their data and your data (replace `\ ` with ` `). applied conditional formatting to show what items from their list match in yours and vice versa (green). left their 'new' column in there so you can see what they recently added to their asset.
![badbots](http://zoggy.net/EXCEL_2017-10-31_15-25-54.png)

excel file: http://zoggy.net/badbots.xlsx

**Comment by thezoggy** [comment]
ok ran latest list through yesterday's apache log, all good:

```
xxx:80 37.115.191.27 - - [30/Oct/2017:06:11:18 +0000] "GET / HTTP/1.1" 200 14600 "http://komputers-best.ru/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.2; Win64; AMD64)"
xxx:80 37.115.191.27 - - [30/Oct/2017:06:11:16 +0000] "GET / HTTP/1.1" 200 14600 "http://komputers-best.ru/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.2; Win64; AMD64)"
xxx:80 138.201.59.34 - - [30/Oct/2017:09:38:38 +0000] "GET /robots.txt HTTP/1.0" 200 271 "-" "Mozilla/5.0 (compatible; SEOkicks-Robot; +http://www.seokicks.de/robot.html)"
xxx:80 71.6.202.198 - - [30/Oct/2017:14:57:32 +0000] "GET /xxbb HTTP/1.1" 404 347 "-" "Mozilla/5.0 zgrab/0.x"
xxx:80 71.6.202.198 - - [30/Oct/2017:15:34:15 +0000] "GET /xxbb HTTP/1.1" 404 347 "-" "Mozilla/5.0 zgrab/0.x"
xxx:80 178.137.177.108 - - [30/Oct/2017:17:52:27 +0000] "GET / HTTP/1.1" 200 14600 "https://ww2awards.info/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows 98; Win 9x 4.90)"
xxx:80 178.137.177.108 - - [30/Oct/2017:17:52:29 +0000] "GET / HTTP/1.1" 200 14600 "https://ww2awards.info/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows 98; Win 9x 4.90)"
xxx:80 178.137.177.108 - - [30/Oct/2017:17:52:30 +0000] "GET / HTTP/1.1" 200 14600 "https://ww2awards.info/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows 98; Win 9x 4.90)"
xxx:80 178.62.99.175 - - [30/Oct/2017:19:09:19 +0000] "GET / HTTP/1.1" 200 19184 "-" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; InfoPath.2)"
xxx:80 188.163.74.19 - - [30/Oct/2017:21:07:14 +0000] "GET / HTTP/1.1" 200 46720 "https://resant.ru/bezopasnoe-ispolzovanie-kart-tinkoff-bank.html" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; .NET CLR 1.1.4322)"
xxx:80 188.163.74.19 - - [30/Oct/2017:21:07:15 +0000] "GET / HTTP/1.1" 200 46720 "https://resant.ru/bezopasnoe-ispolzovanie-kart-tinkoff-bank.html" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; .NET CLR 1.1.4322)"
xxx:80 46.118.159.224 - - [31/Oct/2017:01:11:57 +0000] "GET / HTTP/1.1" 200 14600 "http://ukr-lit.com/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)"
xxx:80 46.118.159.224 - - [31/Oct/2017:01:11:56 +0000] "GET / HTTP/1.1" 200 14600 "http://ukr-lit.com/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)"
xxx:80 46.118.159.224 - - [31/Oct/2017:01:11:59 +0000] "GET / HTTP/1.1" 200 14600 "http://ukr-lit.com/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)"
xxx:80 46.118.159.224 - - [31/Oct/2017:02:42:43 +0000] "GET / HTTP/1.1" 200 14600 "https://colobit.biz/?ref=kissska" "Mozilla/4.0 (compatible; MSIE 6.0; Windows XP)"
xxx:80 46.118.159.224 - - [31/Oct/2017:02:42:42 +0000] "GET / HTTP/1.1" 200 14600 "https://colobit.biz/?ref=kissska" "Mozilla/4.0 (compatible; MSIE 6.0; Windows XP)"
xxx:80 46.118.159.224 - - [31/Oct/2017:02:42:40 +0000] "GET / HTTP/1.1" 200 14600 "http://multgo.ru/" "Mozilla/4.0 (compatible; MSIE 6.0; America Online Browser 1.1; rev1.2; Windows NT 5.1; SV1; .NET CLR 1.1.4322)"
xxx:80 46.118.159.224 - - [31/Oct/2017:02:42:41 +0000] "GET / HTTP/1.1" 200 14600 "http://multgo.ru/" "Mozilla/4.0 (compatible; MSIE 6.0; America Online Browser 1.1; rev1.2; Windows NT 5.1; SV1; .NET CLR 1.1.4322)"
xxx:80 46.118.159.224 - - [31/Oct/2017:02:42:40 +0000] "GET / HTTP/1.1" 200 14600 "https://colobit.biz/?ref=kissska" "Mozilla/4.0 (compatible; MSIE 6.0; Windows XP)"
xxx:80 37.115.189.113 - - [31/Oct/2017:05:33:12 +0000] "GET / HTTP/1.1" 200 14600 "https://landmania.ru/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; Media Center PC"
xxx:80 37.115.189.113 - - [31/Oct/2017:05:33:13 +0000] "GET / HTTP/1.1" 200 14600 "https://landmania.ru/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; Media Center PC"
xxx:80 37.115.189.113 - - [31/Oct/2017:05:33:10 +0000] "GET / HTTP/1.1" 200 14600 "https://landmania.ru/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; Media Center PC"
```

**Comment by thezoggy** [comment]
so I was curious if there is anything in stopbadbots's data that we could use...
so lets use assets/bots.txt and generate an egrep friendly pattern (we will make it use word boundary later):
```
cat stopbadbots/assets/bots.txt | sed -e 's/\([.\:|()+]\)/\\\1/g' | tr '\n' '|' | sed -e 's/|$//g' > ~/stopbadbots.txt
```

then just run the following command to have it see what it would match:
```
egrep -w "`cat stopbadbots.txt`" /var/log/apache2/other_vhosts_access.log
```

snippet:
![badbot2](http://zoggy.net/Xshell_2017-10-31_16-56-34.png)

we already have `zgrab`, `Python-urllib` and `wget` would result in false positives for a lot of people...
but you can see there are two that we dont have, that they do: `SeznamBot` and `CFNetwork`

looking into these two.. one is a czech crawler and cfnetwork is for ios apps... 

http://www.botopedia.org/user-agent-list/search-bots/seznam-bot
https://user-agents.me/cfnetwork-version-list

since I dont really have any ios app users that hit my site I dont know if this would do more harm than good.. looking at my logs I see that all of these cfnetwork hits come from the same two ip's and never actually browser the site.. just hit it for favicon/apple-touch.. so its easy enough for me to say we skip this one and we let the user just block the ip... or we block and people have the option to whitelist later

**Comment by thezoggy** [comment]
added the 6 entries above, reloaded filter and monitoring now. so far so good.

```$ egrep "`sudo fail2ban-client status apache-badbots | grep "Banned IP" | cut -d':' -f 2- | tr ' ' '|'`" /var/log/apache2/other_vhosts_access.log
xxx:80 141.212.122.48 - - [31/Oct/2017:08:42:14 +0000] "GET / HTTP/1.1" 200 19071 "-" "Mozilla/5.0 zgrab/0.x"
xxx:80 104.236.152.56 - - [31/Oct/2017:12:57:57 +0000] "GET / HTTP/1.1" 200 19072 "-" "Mozilla/5.0 zgrab/0.x"
xxx:80 104.236.173.216 - - [31/Oct/2017:14:22:56 +0000] "GET / HTTP/1.1" 200 19073 "-" "Mozilla/5.0 zgrab/0.x"
xxx:80 155.94.88.58 - - [31/Oct/2017:16:03:32 +0000] "GET / HTTP/1.0" 200 25800 "-" "sysscan/1.0 (https://github.com/robertdavidgraham/sysscan)"
```

**Comment by thezoggy** [comment]
btw, if you dont wrap that above content in backticks, it will display as a table

**Comment by mitchellkrogza** [comment]
Thanks @thezoggy fixed it .... looks much better now 👍

**Comment by thezoggy** [comment]
You see my previous note about the 6 entries to add;
`Cloud mapping`, `Baiduspider`, `Who.is Bot`, `SeznamBot`, `evc-batch`, `YandexBot`

**Comment by mitchellkrogza** [comment]
:+1: will add them later today.

**Comment by mitchellkrogza** [comment]
@thezoggy can you maybe send a PR creating a sub folder in the generator lists folder called “fail2ban-bad-bots” with a duplicate of the bad bots list with these additions. Reasons being for the actual bot blocker usage Yandex and Baidu are actually in the blocker but in a rate limited section as some users want to allow access to these bots. So putting them in the bad bots blocked section will not come over well with all the users of this and the Nginx blocker. Then we can maintain a separate list for the fail2ban jail which you and I can curate and moderate.

**Comment by mitchellkrogza** [comment]
@thezoggy Put the bad bost / hosts list for Fail2Ban here with the additions above - https://github.com/mitchellkrogza/apache-ultimate-bad-bot-blocker/tree/master/_fail2ban_bad_hosts

**Comment by thezoggy** [comment]
well we could also just keep the badbots as they are, then a separate list for baidu/yandex/etc which then one can pull in if needed. so for example i would just pull down the badbots then pull down that list and combine.. vs duplicating the whole badbots list again

**Comment by mitchellkrogza** [comment]
👍  ok should I rather move it to the _generator_lists folder and call it 
bad-user-agents-fail2ban-additional.list  ? and then only include the 6 I added ?

**Comment by thezoggy** [comment]
not sure what the 6 you added, but if it's the ones I listed earlier. most of those should be in this list.. and you could segment the baidu/yandex in the legit-but-still-bots list...

`Cloud mapping`, `Who.is Bot`, `SeznamBot`, `evc-batch`

**Comment by mitchellkrogza** [comment]
Ok gotya, main list updated with 4 entries above. Extra's moved to separate list in same folder called `bad-user-agents-fail2ban-additional.list` does this seem like a better solution?

**Comment by mitchellkrogza** [comment]
@thezoggy know anything more about `Spice M5460` my grep results show 808683 results from my logs?

**Comment by thezoggy** [comment]
yes, also we should have a 'test' file with apache data so we can run regex/test against.

so rough draft for me would be like this, goal is to have every line matched

```
sudo fail2ban-regex --print-all-missed sample.log /etc/fail2ban/filter.d/apache-badbots.conf
```

sample.log:
```
localhost:80 178.137.177.108 - - [30/Oct/2017:17:52:30 +0000] "GET / HTTP/1.1" 200 14600 "https://ww2awards.info/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows 98; Win 9x 4.90)"
localhost:80 198.41.89.38 - - [31/Oct/2017:01:48:42 +0000] "GET /apple-touch-icon.png HTTP/1.1" 404 418 "-" "Safari/12604.1.38.1.7 CFNetwork/811.5.4 Darwin/16.7.0 (x86_64)"
localhost:80 198.41.89.38 - - [31/Oct/2017:01:48:42 +0000] "GET /favicon.ico HTTP/1.1" 200 1434 "-" "Safari/12604.1.38.1.7 CFNetwork/811.5.4 Darwin/16.7.0 (x86_64)"
localhost:80 138.201.59.34 - - [31/Oct/2017:10:23:10 +0000] "GET /robots.txt HTTP/1.0" 200 271 "-" "Mozilla/5.0 (compatible; SEOkicks-Robot; +http://www.seokicks.de/robot.html)"
localhost:80 216.244.66.233 - - [31/Oct/2017:13:10:45 +0000] "GET /robots.txt HTTP/1.1" 200 252 "-" "Mozilla/5.0 (compatible; DotBot/1.1; http://www.opensiteexplorer.org/dotbot, help@moz.com)"
localhost:80 104.236.173.216 - - [31/Oct/2017:14:22:56 +0000] "GET / HTTP/1.1" 200 19073 "-" "Mozilla/5.0 zgrab/0.x"
localhost:80 155.94.88.58 - - [31/Oct/2017:16:03:32 +0000] "GET / HTTP/1.0" 200 25800 "-" "sysscan/1.0 (https://github.com/robertdavidgraham/sysscan)"
localhost:80 37.115.191.49 - - [01/Nov/2017:12:29:05 +0000] "GET / HTTP/1.1" 200 43800 "http://03p.info/" "Mozilla/4.7 (compatible; OffByOne; Windows 2000) Webster Pro V3.4"
localhost:80 37.115.191.49 - - [01/Nov/2017:12:29:05 +0000] "GET / HTTP/1.1" 200 40880 "https://ilmexico.com/" "Mozilla/4.7 (compatible; OffByOne; Windows 2000) Webster Pro V3.4"
localhost:80 123.249.3.172 - - [01/Nov/2017:12:36:46 +0000] "GET / HTTP/1.1" 200 412 "-" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.0)"
localhost:80 35.193.153.123 - - [01/Nov/2017:15:16:04 +0000] "GET /robots.txt HTTP/1.1" 200 271 "-" "ltx71 - (http://ltx71.com/)"
localhost:80 35.193.153.123 - - [01/Nov/2017:15:16:34 +0000] "GET / HTTP/1.1" 200 19149 "-" "ltx71 - (http://ltx71.com/)"
localhost:80 37.115.191.49 - - [02/Nov/2017:03:07:01 +0000] "GET / HTTP/1.1" 200 14600 "http://astana.localhostkaz.org/" "Mozilla/4.0 (compatible; MSIE 5.5; Windows NT 4.0; .NET CLR 1.0.2914)"
localhost:80 170.79.35.196 - - [02/Nov/2017:12:09:38 +0000] "HEAD http://23.94.61.110:80/mysql/admin/ HTTP/1.1" 404 609 "-" "Mozilla/5.0 Jorgee"
```

looking at the results, I see that there are a few that arent being caught:
```
Lines: 14 lines, 0 ignored, 11 matched, 3 missed
[processed in 0.54 sec]

|- Missed line(s):
|  localhost:80 198.41.89.38 - - [31/Oct/2017:01:48:42 +0000] "GET /apple-touch-icon.png HTTP/1.1" 404 418 "-" "Safari/12604.1.38.1.7 CFNetwork/811.5.4 Darwin/16.7.0 (x86_64)"
|  localhost:80 198.41.89.38 - - [31/Oct/2017:01:48:42 +0000] "GET /favicon.ico HTTP/1.1" 200 1434 "-" "Safari/12604.1.38.1.7 CFNetwork/811.5.4 Darwin/16.7.0 (x86_64)"
|  localhost:80 37.115.191.49 - - [02/Nov/2017:03:07:01 +0000] "GET / HTTP/1.1" 200 14600 "http://astana.localhostkaz.org/" "Mozilla/4.0 (compatible; MSIE 5.5; Windows NT 4.0; .NET CLR 1.0.2914)"
`-
```

etc.. that way I know if I need to update regex.. or if an update broke something.

**Comment by thezoggy** [comment]
I do not know about `Spice M5460`, looking for `\bSpice\b` in my logs does not show any results.
looking online I see a few Spice variants about being a data miner but most of these sites are 'generated' so I dont know if I can trust it. Can you post the apache log so I can see the type of request/user agent string

**Comment by thezoggy** [comment]
looking through recent logs, one thing that I noticed is that how a lot of fake user agents use the `MSIE 6.0` string which you already block.. but there is a bunch that use `Mozilla/4.##`

go run this on your sites:
`egrep -iw "Mozilla/4" apache.log`

(you can see that \bMozilla/4\b is what it basically is doing and how it matches)..

```
$ egrep -iw "Mozilla/4" apache.log 
localhost:80 80.40.134.104 - - [03/Nov/2017:06:33:50 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 62.24.196.100 - - [03/Nov/2017:07:22:45 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 62.24.252.133 - - [03/Nov/2017:07:43:17 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 80.40.134.103 - - [03/Nov/2017:10:01:07 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 188.163.75.12 - - [03/Nov/2017:10:38:41 +0000] "GET / HTTP/1.1" 200 43800 "https://schoolfiles.net/" "Mozilla/4.7 (compatible; OffByOne; Windows 2000) Webster Pro V3.4"
localhost:80 188.163.75.12 - - [03/Nov/2017:10:38:42 +0000] "GET / HTTP/1.1" 200 58400 "https://schoolfiles.net/" "Mozilla/4.7 (compatible; OffByOne; Windows 2000) Webster Pro V3.4"
localhost:80 188.163.75.12 - - [03/Nov/2017:11:02:53 +0000] "GET / HTTP/1.1" 200 105120 "https://resant.ru/bezopasnoe-ispolzovanie-kart-tinkoff-bank.html" "Mozilla/4.0 (compatible; MSIE 5.0; Windows NT; DigExt)"
localhost:80 188.163.75.12 - - [03/Nov/2017:11:02:54 +0000] "GET / HTTP/1.1" 200 78840 "https://resant.ru/bezopasnoe-ispolzovanie-kart-tinkoff-bank.html" "Mozilla/4.0 (compatible; MSIE 5.0; Windows NT; DigExt)"
localhost:80 46.118.153.154 - - [03/Nov/2017:13:05:46 +0000] "GET / HTTP/1.1" 200 14600 "http://choice-credit.ru/obzory/557-polozhitelnye-svojstva-kreditnykh-kart" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; Media Center PC"
localhost:80 46.118.153.154 - - [03/Nov/2017:13:05:49 +0000] "GET / HTTP/1.1" 200 14600 "http://choice-credit.ru/obzory/557-polozhitelnye-svojstva-kreditnykh-kart" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; Media Center PC"
localhost:80 46.118.153.154 - - [03/Nov/2017:13:05:51 +0000] "GET / HTTP/1.1" 200 14600 "http://choice-credit.ru/obzory/557-polozhitelnye-svojstva-kreditnykh-kart" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; Media Center PC"
localhost:80 62.24.196.100 - - [03/Nov/2017:13:59:29 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 211.232.173.36 - - [03/Nov/2017:14:11:55 +0000] "GET / HTTP/1.1" 200 19163 "-" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1)"
localhost:80 211.232.173.36 - - [03/Nov/2017:15:07:29 +0000] "GET / HTTP/1.1" 200 19238 "-" "Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 6.1; WOW64; Trident/5.0; SLCC2; .NET CLR 2.0.50727; .NET CLR 3.5.30729; .NET CLR 3.0.30729; Media Center PC 6.0; .NET4.0C; InfoPath.3; .NET4.0E)"
localhost:80 37.115.191.49 - - [03/Nov/2017:15:08:24 +0000] "GET / HTTP/1.1" 200 39420 "http://14b.info/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; YPC 3.0.2; .NET CLR 1.1.4322; yplus 4.4.02b)"
localhost:80 37.115.191.49 - - [03/Nov/2017:15:08:24 +0000] "GET / HTTP/1.1" 200 14600 "https://ilmexico.com/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; MyIE2; Deepnet Explorer)"
localhost:80 37.115.191.49 - - [03/Nov/2017:15:08:25 +0000] "GET / HTTP/1.1" 200 39420 "http://14b.info/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; YPC 3.0.2; .NET CLR 1.1.4322; yplus 4.4.02b)"
localhost:80 37.115.191.49 - - [03/Nov/2017:15:08:25 +0000] "GET / HTTP/1.1" 200 39420 "https://ilmexico.com/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; MyIE2; Deepnet Explorer)"
localhost:80 37.115.191.49 - - [03/Nov/2017:15:08:26 +0000] "GET / HTTP/1.1" 200 37960 "http://14b.info/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; YPC 3.0.2; .NET CLR 1.1.4322; yplus 4.4.02b)"
localhost:80 37.115.191.49 - - [03/Nov/2017:15:08:26 +0000] "GET / HTTP/1.1" 200 40880 "https://ilmexico.com/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; MyIE2; Deepnet Explorer)"
localhost:80 211.232.159.3 - - [03/Nov/2017:15:50:51 +0000] "GET / HTTP/1.1" 200 19162 "-" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1)"
localhost:80 50.112.194.65 - - [03/Nov/2017:15:56:12 +0000] "GET / HTTP/1.1" 200 184254 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)"
localhost:80 45.55.224.77 - - [03/Nov/2017:15:58:08 +0000] "GET / HTTP/1.1" 200 19174 "-" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; InfoPath.2)"
localhost:80 211.232.159.3 - - [03/Nov/2017:17:04:56 +0000] "GET / HTTP/1.1" 200 19149 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; InfoPath.2)"
localhost:80 211.232.173.34 - - [03/Nov/2017:18:16:12 +0000] "GET / HTTP/1.1" 200 19149 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; InfoPath.2)"
localhost:80 211.232.159.2 - - [03/Nov/2017:18:36:21 +0000] "GET / HTTP/1.1" 200 19124 "-" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1)"
localhost:80 62.24.222.132 - - [03/Nov/2017:19:01:13 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 211.232.173.34 - - [03/Nov/2017:19:20:16 +0000] "GET / HTTP/1.1" 200 19149 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; InfoPath.2)"
```

I do not see it doing any false positives from my data.
Since the fake agent can be safely blocked since the agent is so old now.


I dont know of any browsers or legit traffic that queries robots.txt except for crawlers/bots/nasty things..  
I can see that the fake agent is heavily used there:
```
$ grep robots.txt apache.log | grep Mozilla
localhost:80 80.40.134.104 - - [03/Nov/2017:06:33:50 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 62.24.196.100 - - [03/Nov/2017:07:22:45 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 62.24.252.133 - - [03/Nov/2017:07:43:17 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 80.40.134.103 - - [03/Nov/2017:10:01:07 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 138.201.59.34 - - [03/Nov/2017:12:03:47 +0000] "GET /robots.txt HTTP/1.0" 200 271 "-" "Mozilla/5.0 (compatible; SEOkicks-Robot; +http://www.seokicks.de/robot.html)"
localhost:80 142.54.187.51 - - [03/Nov/2017:12:24:41 +0000] "GET /robots.txt HTTP/1.1" 200 271 "-" "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.80 Safari/537.36"
localhost:80 62.24.196.100 - - [03/Nov/2017:13:59:29 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 62.24.222.132 - - [03/Nov/2017:19:01:13 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 62.24.252.133 - - [03/Nov/2017:19:49:34 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 62.24.252.133 - - [03/Nov/2017:21:45:53 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 62.24.181.135 - - [03/Nov/2017:23:03:56 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
localhost:80 80.40.134.104 - - [03/Nov/2017:23:22:22 +0000] "GET /robots.txt HTTP/1.0" 200 271 "http://localhost/robots.txt" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; .NET CLR 2.0.50727; .NET CLR 3.0.04506.648; .NET CLR 3.5.21022; .NET CLR 3.0.4506.2152; .NET CLR 3.5.30729; InfoPath.2)"
```

so I say we add:
`Mozilla/4`

**Comment by thezoggy** [comment]
fyi in your list you have:
```
Who.is
Who.is\ Bot
```

the first takes care of the 2nd..  is there a reason your adding the first variant?

**Comment by thezoggy** [comment]
Saw this one, 
`localhost:80 85.248.227.165 - - [03/Nov/2017:22:43:47 +0000] "GET / HTTP/1.1" 200 19087 "AfD-Verbotsverfahren_JETZT!" "AfD-Verbotsverfahren JETZT!"`

should add:
`AfD-Verbotsverfahren`

**Comment by thezoggy** [comment]
to circle back on, `CFNetwork` that wont work because the boundary gets in the way of the suffix.. so leave that one off.. was easy enough to just block ip for those offenders.

**Comment by mitchellkrogza** [comment]
Hi @thezoggy hope you are well, sorry I've been ill for the past 10 days and only now surfacing again. 

I have addressed the `Who.is bot` issue and added `AfD-Verbotsverfahren` ... not sure about `Mozilla/4` have you given more thought to this?

**Comment by thezoggy** [comment]
You currently have `MSIE 6.0`, which blocks out a lot of spam/bad things because its so old and the likely hood of someone actually using IE6 is so low. The only things that actively are using it are bots/fake agents.

`Mozilla/4` imho is basically in the same spot but just slightly newer as WinXP IE8 would be affected.

> Support for all versions of IE on Windows XP ended effective April 8, 2014 due to its end-of-life. On January 12, 2016, support for IE8 on all supported Windows operating systems ceased.

Which is why I personally have blocked it on my side and seen it block out several spam bots because of it... 
Why I'm curious if you could run it against your site and see if you feel the same.

some example where `Mozilla/4` have been effective.. can see that `MSIE 6.0` is in there with them a bunch..
```
localhost:80 178.137.177.108 - - [30/Oct/2017:17:52:30 +0000] "GET / HTTP/1.1" 200 14600 "https://ww2awards.info/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows 98; Win 9x 4.90)"
localhost:80 37.115.191.49 - - [01/Nov/2017:12:29:05 +0000] "GET / HTTP/1.1" 200 43800 "http://03p.info/" "Mozilla/4.7 (compatible; OffByOne; Windows 2000) Webster Pro V3.4"
localhost:80 37.115.191.49 - - [01/Nov/2017:12:29:05 +0000] "GET / HTTP/1.1" 200 40880 "https://ilmexico.com/" "Mozilla/4.7 (compatible; OffByOne; Windows 2000) Webster Pro V3.4"
localhost:80 123.249.3.172 - - [01/Nov/2017:12:36:46 +0000] "GET / HTTP/1.1" 200 412 "-" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.0)"
localhost:80 37.115.191.49 - - [02/Nov/2017:03:07:01 +0000] "GET / HTTP/1.1" 200 14600 "http://astana.localhostkaz.org/" "Mozilla/4.0 (compatible; MSIE 5.5; Windows NT 4.0; .NET CLR 1.0.2914)"
localhost:80 46.118.153.154 - - [03/Nov/2017:13:05:46 +0000] "GET / HTTP/1.1" 200 14600 "http://choice-credit.ru/obzory/557-polozhitelnye-svojstva-kreditnykh-kart" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; Media Center PC"
localhost:80 37.115.191.49 - - [03/Nov/2017:15:08:24 +0000] "GET / HTTP/1.1" 200 39420 "http://14b.info/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; YPC 3.0.2; .NET CLR 1.1.4322; yplus 4.4.02b)"
localhost:80 45.55.224.77 - - [03/Nov/2017:15:58:08 +0000] "GET / HTTP/1.1" 200 19174 "-" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; InfoPath.2)"
localhost:80 46.118.118.35 - - [03/Nov/2017:22:47:52 +0000] "GET / HTTP/1.1" 200 14600 "https://liky.co.ua/" "Mozilla/4.76 [en] (Windows NT 5.0; U)"
localhost:80 46.118.153.154 - - [10/Nov/2017:11:32:41 +0000] "GET / HTTP/1.1" 200 14600 "http://multgo.ru/" "Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1; SV1; .NET CLR 1.1.4322) Babya Discoverer  8.0:"
```

**Comment by thezoggy** [comment]
so I use your list and put a word boundary on both ends.. since thats how you said above thats how you use it.. 

so for example,
`Netcraft` becomes `\bNetcraft\b`

which means it would NOT catch:
```
localhost:80 138.197.11.229 - - [11/Nov/2017:06:48:59 +0000] "GET / HTTP/1.1" 200 19183 "-" "Mozilla/5.0 (compatible; NetcraftSurveyAgent/1.0; +info@netcraft.com)"
```

just like `Zgrab` -> `\bZgrab\b` doesnt work for:
```
localhost:80 104.236.173.216 - - [31/Oct/2017:14:22:56 +0000] "GET / HTTP/1.1" 200 19073 "-" "Mozilla/5.0 zgrab/0.x"
```

so, if this is the case then we need to update the list or the regex so they actually match.
thoughts?

**Comment by mitchellkrogza** [comment]
@thezoggy `Mozilla/4` now integrated in https://github.com/mitchellkrogza/apache-ultimate-bad-bot-blocker/commit/9078d6c4ca4f09cb9cf0d7f8c5ad44903419803e

Would a single word boundary not work better? `\bNetcraft` instead of `\bNetcraft\b` ??

**Comment by thezoggy** [comment]
I'm open to whatever.. I was just trying to match what you do (so the list is valid).. and I noticed the list isnt valid if your doing a boundary on both ends..  if you dont do one at the end.. I wonder if there would be any false positives... because you have some really vague words and some two letter ones.. which scares me. It seems like we should just update the entries to make it work with a leading and ending boundary?

so `Zgrab` should be `zgrab/0`
and `Netcraft` should be `NetcraftSurveyAgent/1`

?

**Comment by mitchellkrogza** [comment]
@thezoggy I think keeping the list updated with the correct names for the bots is a much better idea and will lead to much less false positives, if any. Will action those 2 for now.

**Comment by thezoggy** [comment]
so you may not have to change anything.. per  fail2ban-regex i see it does match with the way it is right now

**Comment by mitchellkrogza** [comment]
@thezoggy so should `NetcraftSurveyAgent/1` and `zgrab/0` stay as is or should we also have `zgrab` and `Netcraft` in the list? My brain is not functioning properly today 😴

**Comment by thezoggy** [comment]
I see why I'm matching.. in my regex I'm greedy.. so even though my terms have \b wrapped around them.. I'm matching them in a .* block.. so the \b actually aren't being applied...
If I 'fix' my regex then I see it not match as I saw with the egrep -iw quick test.

**Comment by thezoggy** [comment]
When your project tests against the user agent.. is it finding the match anywhere in the log line? or is it processing only an extracted string from it?

**Comment by mitchellkrogza** [comment]
My tests have always been like this 

`curl -A "Xenu Link Sleuth/1.3.8" http://yourdomain.com`

and

`curl -A "Mozilla/5.0 (compatible; AhrefsBot/5.2; +http://ahrefs.com/robot/)" http://yourdomain.com` for example

**Comment by thezoggy** [comment]
right, but your logic must actually be matching the block anywhere in the agent.. not anchored at the start or end.. lets say you have `AhrefsBot`, and you do boundary on both ends.. `\bAhrefsBot\b` then you can see that it wouldnt actually match.. the space leading is fine.. but the trailing slash would break it.. which makes me think you must have it matching .* before or after..  for the same reasons I do.. so you can match the term anywhere and not anchor it.. 

if you anchored it to the start you'd have to put `"Mozilla/5.0 (compatible; AhrefsBot`
and if you anchored it at the end... youd have to do `AhrefsBot/5.2; +http://ahrefs.com/robot/)"`

which means I bet your word boundary arent really doing you any good either..

**Comment by mitchellkrogza** [comment]
Okay I see why I am getting a match for Ahrefs, look at Line 223 and Line 1155 of https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/blob/master/conf.d/globalblacklist.conf ... it's getting matched on the referrer and not the bot name.

**Comment by thezoggy** [comment]
anyways the list is fine with me now that I understand how its matching.. you can revert it back to zgrab / netcraft. because netcraft doesnt match in NetcraftSurveyAgent because of the trailing \b but it does match in the latter part of the agent.. netcraft.com so it works.

**Comment by mitchellkrogza** [comment]
Great @thezoggy will do and thanks for your input into this 👍

**Comment by thezoggy** [comment]
to start the conversation, what is a bad bot. 

In your list you have `YoudaoBot` which is a Chinese search engine but you put Baidu in a different list.
in my eyes they are both bad because i do not serve up a site to anyone in china.. and/or i dont want any public search engine scraping my site. Now if I was running a eCommerce site that was global.. I prob wouldnt be blocking anything (loss revenue).

So now the scope of usa (domestic) or global comes into play. Is google okay but china sites arent? etc.

so anyways, food for thought
----------------------------------------------------------------

Looking at this url, its data is from 2010 (so some of the examples may not really be valid anymore) but you can get the idea:
https://searchenginewatch.com/sew/news/2067357/bye-bye-crawler-blocking-parasites

looks like we could add: `Sogou`, `NaverBot`

and about:
https://github.com/mitchellkrogza/apache-ultimate-bad-bot-blocker/blob/master/_generator_lists/bad-user-agents-fail2ban-additional.list

all those yandex variants cant just be simplified to `yandex` because it will match in the later part of the agent.. `yandex.com`

**Comment by mitchellkrogza** [comment]
Thank @thezoggy will give it some thought. 

The blocker was initially developed to suit my own needs and I don't really sell to certain markets who merely infringe my copyrights, scrape my content and such. 

I added `Mozilla/4.0` in addition to `Mozilla/4` this morning as the `/4.0` was not being caught due to the word boundary. 

Will take a look and sort out the Yandex variants 👍

**Comment by mitchellkrogza** [comment]
@thezoggy as for the bad-user-agents-fail2ban-additional.list

```
Baidu
Baiduspider
Yandex
YandexBot
YandexImages
```

Removing `YandexBot` and `YandexImages` and leaving only `Yandex` will not result in `Yandex` matching `YandexBot` due to the word boundary.

Would it not be better for me to amend the regex catch to only a beginning word boundary? or do you think we risk false positives again? Trying to find some stable ground here

**Comment by thezoggy** [comment]
you are correct Yandex would not match YandexBot. but thats not what im saying it will match..  

per: https://yandex.com/support/webmaster/robot-workings/check-yandex-robots.xml
see how they all have `+http://yandex.com/bots` in the agent..

thats why just Yandex works, because it will match in the yandex.com part.
and I dont think we should do only one boundary.. would be way to easy to introduce false positives..

**Comment by thezoggy** [comment]
about `Mozilla/4`, the `.` is part of \b.. so if its not working then something else is going on... it works for me

**Comment by mitchellkrogza** [comment]
Thanks @thezoggy I'll check through things and see if I can trace the Mozilla/4 issue

**Comment by mitchellkrogza** [comment]
@thezoggy complaints about Mozilla/4 and MSIE 6.0 https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/issues/110 so removed and moved them to https://github.com/mitchellkrogza/apache-ultimate-bad-bot-blocker/blob/master/_generator_lists/bad-user-agents-fail2ban-additional.list

**Comment by thezoggy** [comment]
ok, lets drop `Mozilla/4` and `Mozilla/4.0` from the additional list. 
lets add to that list however, `bingbot`:
```
localhost:80 207.46.13.197 - - [17/Nov/2017:04:44:49 +0000] "GET /robots.txt HTTP/1.1" 200 308 "-" "Mozilla/5.0 (compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm)"
```

**Comment by thezoggy** [comment]
looks like we should add `ExtLinksBot` to the main list:
```
localhost:80 142.4.200.37 - - [15/Nov/2017:17:24:28 +0000] "GET /robots.txt HTTP/1.1" 200 308 "-" "Mozilla/5.0 (compatible; ExtLinksBot/1.5 +https://extlinks.com/Bot.html)"
```

**Comment by thezoggy** [comment]
also need to add `Uptimebot` to the main list:
```
localhost:80 74.115.214.157 - - [08/Nov/2017:08:43:28 +0000] "GET /robots.txt HTTP/1.1" 200 271 "-" "Mozilla/5.0 (compatible; Uptimebot/1.0; +http://www.uptime.com/uptimebot)"

**Comment by mitchellkrogza** [comment]
@thezoggy  `UpTimeBot` used to be on the main list but was removed as some users use their service. `applebot`, `Yahoo\! Slurp` might also interfere with some users so these 3 will have to go on the additional list to avoid interfering with the blocker on web servers. I will check my logs for `oBot`, `filterdb` and the the other 2

**Comment by mitchellkrogza** [comment]
Here's an interesting new one spotted in one of my log files today `PHP/{5|6|7}.{3|2}.{1|2|3|4|5|6|7|8|9|0}{1|2|3|4|5|6|7|8|9|0}`

**Comment by thezoggy** [comment]
can you post an example? (that regex could be better written like `PHP/[5-7].[2-3].[0-9][0-9]` or even better.. `PHP/[5-7].[2-3].\d\d`

I wonder if the 'numeric' ones that I've seen. I have a `apache-numeric` that handles those...
`failregex = <HOST> -.*"(GET|POST|HEAD) http\:\/\/\d+\.\d+\.\d+\.\d+.*$`

**Comment by mitchellkrogza** [comment]
@thezoggy Here's a log excerpt of that one.

`36.63.30.241 - - [21/Nov/2017:18:41:45 +0200] "GET / HTTP/1.1" 200 7830 "-" "PHP/{5|6|7}.{3|2}.{1|2|3|4|5|6|7|8|9|0}{1|2|3|4|5|6|7|8|9|0}" "-"PORT:443 0.000 - . "GZIP:4.05"
`

**Comment by thezoggy** [comment]
since your not blocking mozilla/4, how do you feel about MSIE 5.0 or Windows 3.1? 
```
localhost:80 46.118.119.80 - - [22/Nov/2017:12:06:19 +0000] "GET / HTTP/1.1" 200 14600 "http://www.szucs.ru/" "Mozilla/4.0 (compatible; MSIE 5.0; Windows 3.1)"
```
https://translate.google.com/translate?hl=en&sl=ru&u=http://www.szucs.ru/&prev=search

**Comment by thezoggy** [comment]
@mitchellkrogza thats an odd one.. looks like a script gone bad..

**Comment by thezoggy** [comment]
Per: https://www.bleepingcomputer.com/news/security/theres-some-intense-web-scans-going-on-for-bitcoin-and-ethereum-wallets/

Looks like `Firefox/21.0` may be a good one to block.. browser is so old there isnt any harm?

**Comment by thezoggy** [comment]
more bots to block:

```localhost:80 34.192.116.178 - - [18/Oct/2017:14:17:02 +0000] "GET /robots.txt HTTP/1.1" 200 308 "-" "Mozilla/5.0 (compatible; SMTBot/1.0; +http://www.similartech.com/smtbot)"
localhost:80 144.202.193.40 - - [19/Oct/2017:03:27:29 +0000] "GET /robots.txt HTTP/1.1" 200 271 "-" "Mozilla/5.0 (compatible; FemtosearchBot/1.0; http://femtosearch.com)"
localhost:80 146.185.130.221 - - [19/Nov/2017:17:11:46 +0000] "GET /robots.txt HTTP/1.0" 200 271 "-" "SafeDNSBot (https://www.safedns.com/searchbot)"
localhost:80 45.77.209.100 - - [22/Nov/2017:13:14:44 +0000] "GET / HTTP/1.1" 200 184855 "-" "Mozilla/5.0 (compatible; Nmap Scripting Engine; https://nmap.org/book/nse.html)"
localhost:80 45.77.209.100 - - [22/Nov/2017:13:14:45 +0000] "OPTIONS / HTTP/1.1" 200 15928 "-" "Mozilla/5.0 (compatible; Nmap Scripting Engine; https://nmap.org/book/nse.html)"
localhost:80 69.58.178.58 - - [02/Dec/2017:16:29:36 +0000] "GET /robots.txt HTTP/1.1" 200 308 "-" "Mozilla/5.0 (compatible; ips-agent)"
```

something else to add to the additional list,
internet archive (waybackmachine):
```
localhost:80 207.241.229.236 - - [23/Oct/2017:11:24:13 +0000] "GET /robots.txt HTTP/1.0" 200 271 "-" "Mozilla/5.0 (compatible; archive.org_bot +http://www.archive.org/details/archive.org_bot)"
localhost:80 34.207.98.73 - - [25/Oct/2017:04:09:08 +0000] "GET /robots.txt HTTP/1.0" 200 271 "-" "ia_archiver"
```

**Comment by corbolais** [comment]
hey,
thanks for your work. Great job.
I'm considering deploying aubbb. One more bb to block would be "shodan", the vulnerability scanner.
Are you inclined to include it in the list?

**Comment by mitchellkrogza** [comment]
@thezoggy apologies for delay in responding been really under the weather for some time but getting back into the swing of things again. I will address these additions shortly.

@corbolais I am sure we can look into adding shodan, will need to do some checking on it first but you also have full control yourself with the include files to block any additional bots that are not on the main list.

**Comment by thezoggy** [comment]
submitted pull request to get everything up to sync with my related comments

**Comment by thezoggy** [comment]
```
[30/Jan/2018:18:13:27 +0000] "GET / HTTP/1.0" 200 32032 "-" "muhstik-scan"
[30/Jan/2018:18:13:25 +0000] "GET /webdav HTTP/1.0" 404 1768 "-" "muhstik-scan/1.0"
```

looks like a new one to block

**Comment by mitchellkrogza** [comment]
Thanks @thezoggy I will push this out in today's update. 👍 Thanks very much for the contributions

**Comment by thezoggy** [comment]
new one popped up today:
```localhost:80 5.188.9.80 - - [19/Mar/2018:10:02:39 +0000] "GET /wp-config.php~ HTTP/1.1" 404 1853 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.9.80 - - [19/Mar/2018:10:02:39 +0000] "GET /wp/wp-config.php.bak HTTP/1.1" 404 1854 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.9.80 - - [19/Mar/2018:10:02:40 +0000] "GET /wp/.wp-config.php.swp HTTP/1.1" 404 1853 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.9.80 - - [19/Mar/2018:10:02:40 +0000] "GET /wp/wp-config.php~ HTTP/1.1" 404 1853 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.9.80 - - [19/Mar/2018:10:02:40 +0000] "GET /wordpress/wp-config.php.bak HTTP/1.1" 404 1854 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.9.80 - - [19/Mar/2018:10:02:40 +0000] "GET /wordpress/.wp-config.php.swp HTTP/1.1" 404 1853 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.9.80 - - [19/Mar/2018:10:02:40 +0000] "GET /wordpress/wp-config.php~ HTTP/1.1" 404 1853 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.9.80 - - [19/Mar/2018:10:02:41 +0000] "GET /blog/wp-config.php.bak HTTP/1.1" 404 1854 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.9.80 - - [19/Mar/2018:10:02:41 +0000] "GET /blog/.wp-config.php.swp HTTP/1.1" 404 1854 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.9.80 - - [19/Mar/2018:10:02:41 +0000] "GET /blog/wp-config.php~ HTTP/1.1" 404 1853 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.203.40 - - [19/Mar/2018:12:13:51 +0000] "GET /wp-login.php HTTP/1.1" 404 1856 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.203.40 - - [19/Mar/2018:12:13:51 +0000] "GET /wp/wp-login.php HTTP/1.1" 404 1856 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.203.40 - - [19/Mar/2018:12:13:52 +0000] "GET /wordpress/wp-login.php HTTP/1.1" 404 1856 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.203.40 - - [19/Mar/2018:12:13:52 +0000] "GET /blog/wp-login.php HTTP/1.1" 404 1856 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.203.40 - - [19/Mar/2018:12:13:52 +0000] "GET /wp-admin/install.php HTTP/1.1" 404 1855 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.203.40 - - [19/Mar/2018:12:13:52 +0000] "GET /wp/wp-admin/install.php HTTP/1.1" 404 1855 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.203.40 - - [19/Mar/2018:12:13:52 +0000] "GET /wordpress/wp-admin/install.php HTTP/1.1" 404 1855 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.203.40 - - [19/Mar/2018:12:13:52 +0000] "GET /blog/wp-admin/install.php HTTP/1.1" 404 1855 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.203.40 - - [19/Mar/2018:12:13:53 +0000] "GET /wp-config.php.bak HTTP/1.1" 404 1855 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.203.40 - - [19/Mar/2018:12:13:53 +0000] "GET /.wp-config.php.swp HTTP/1.1" 404 1855 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
localhost:80 5.188.203.40 - - [19/Mar/2018:12:13:53 +0000] "GET /wp-config.php~ HTTP/1.1" 404 1855 "-" "NYU Internet Census (https://scan.lol; research@scan.lol)"
```

**Comment by mitchellkrogza** [comment]
I saw this too did a look into them and they look perfectly  harmless ..... For now

**Comment by mitchellkrogza** [comment]
@thezoggy any input on these for me? https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker/issues/143

**Comment by thezoggy** [comment]
let me take a look

**Comment by mitchellkrogza** [comment]
@thezoggy what's your opinion on `shodan` as mentioned by @corbolais

**Comment by corbolais** [comment]
You got me confused here..
Why do you even have to ask this question?
You never heard of shodan before?
Shodan is **the** sw scanner used for vulnerability exploits out there.
Claiming a security focus on networks, intrusion detection and safer/cleaner internets, w/o knowledge about the general idea what shodan is all about seems -- _uh_ -- surprising?

**Comment by mitchellkrogza** [comment]
Excuse me @corbolais I do actually happen to know what shodan is and don;t appreciate the tone of your message, that's not the question being asked here. The question is whether other people are picking up scans of their servers / sites by shodan. Anyone can use a vulnerability scanner to scan their OWN servers / sites but when someone else takes a vuln / pen testing tool and points it at servers that do not belong to them then .... that's what we call a BAD BOT

**Comment by wumpus** [comment]
Hi, I'm the developer of glindahl-cocrawler, and yeah, stopbadbots lists everything they see as a bad bot, no curation, no appeal.

Also, it's worth noting that DuckDuckGo delivers a lot of results from Yandex, so you might not want to block the Yandex crawler if you want traffic from DuckDuckGo users.

**Comment by mitchellkrogza** [comment]
Hi @wumpus Yandex is not blocked by this blocker, it's specified in a rate limiting section which for the Apache Blocker is not yet active meaning currently they are allowed through but on the Nginx blocker they are rate limited if their crawls get too aggressive.

**Comment by mitchellkrogza** [comment]
This topic has grown too long, closing and reopening a new issue.
